### PR TITLE
[ActionSheet] Add support dynamic type using UIFont scaled fonts

### LINKED ISSUE #8622: [ActionSheet] Internal issue: b/143145691
This was filed as an internal issue. If you are a Googler, please visit [b/143145691](http://b/143145691) for more details.

---
#### Internal data
- Associated internal bug: [b/143145691](http://b/143145691)

### PR DESCRIPTION
**Pull request by codeman7**
## Context
Adds support for UIFont's preferredFont... APIs to render Dynamic Type
fonts. Adds conformance to UIContentSizeCategoryAdjusting so that the fonts
can be adjusted automatically if desired by clients.

## Testing
1. Open up MDCDragons app
2. Open the ActionSheet -> "Material : UIKit comparison" example
3. Change the system font size within settings application

Additionally snapshot test have been added.

Closes #8622

### CODE REVIEW

File: components/ActionSheet/examples/ActionSheetComparisonExampleViewController.m [hunk 1]
```diff
@@ -106,6 +106,17 @@ - (void)showMaterialActionSheet {
   [actionSheet addAction:favoriteAction];
   [actionSheet addAction:emailAction];
   [actionSheet applyThemeWithScheme:self.containerScheme];
+
+  UIFont *font = [UIFont fontWithName:@"Zapfino" size:8];
```
Comment (romoore) [inline components/ActionSheet/examples/ActionSheetComparisonExampleViewController.m]: Were these left in by mistake? I don't believe the examples should demonstrate using Zapfino and we can just stick with the current examples.  Snapshot tests should be sufficient to verify the UIFontMetrics support.

File: components/ActionSheet/src/MDCActionSheetController.m [hunk 2]
```diff
@@ -177,17 +179,18 @@ - (void)viewDidLoad {
   [self.view addSubview:self.headerDividerView];
 }
 
-- (void)viewWillLayoutSubviews {
-  [super viewWillLayoutSubviews];
+- (void)viewDidLayoutSubviews {
```
Comment (romoore) [inline components/ActionSheet/src/MDCActionSheetController.m:182]: Please split this change into a separate PR. It looks like it fixes a bug detectable with the current snapshot tests, so it should be pretty easy to split out and submit.  Then the PR for UIFontMetrics can be submitted in isolation.  In case either of these changes causes regressions internally, we won't lose the other if we  roll-back.

File: components/ActionSheet/src/private/MDCActionSheetHeaderView.h [hunk 3]
```diff
@@ -36,4 +36,10 @@
 
 @property(nonatomic, strong, nullable) UIColor *messageTextColor;
 
+/** Shows the title of the Action Sheet. */
+@property(nonatomic, nonnull, strong, readonly) UILabel *titleLabel;
```
Comment (romoore) [inline components/ActionSheet/src/private/MDCActionSheetHeaderView.h]: If you're making this change, please remove the `font` and `color` properties and just assign them directly to the labels' properties.

File: components/ActionSheet/src/private/MDCActionSheetItemTableViewCell.h [hunk 4]
```diff
@@ -31,6 +31,9 @@
 
 @property(nonatomic, strong, nullable) UIColor *actionTextColor;
 
+/** Label for the action's title text. */
+@property(nonatomic, nonnull, readonly, strong) UILabel *actionLabel;
```
Comment (romoore) [inline components/ActionSheet/src/private/MDCActionSheetItemTableViewCell.h]: Same request here - now that the label is exposed, please remove the properties that passed-through to the label.

File: components/ActionSheet/tests/snapshot/MDCActionSheetControllerSnapshotTests.m [hunk 5]
```diff
@@ -968,4 +984,117 @@ - (void)testActionSheetWithHeaderShownButNoTitleOrMessage {
...
+                               UIContentSizeCategoryAccessibilityExtraExtraExtraLarge];
+    [window traitCollectionDidChange:nil];
+    [self.actionSheetController.view layoutIfNeeded];
+    window.bounds = CGRectMake(0, 0, 1000, 1000);
+    self.actionSheetController.view.frame = window.bounds;
+
+    // Then
+    // Can't add a UIWindow to a UIView, so just screenshot the window directly.
+    [window layoutIfNeeded];
+    [self snapshotVerifyView:window];
```
Comment (romoore) [inline components/ActionSheet/tests/snapshot/MDCActionSheetControllerSnapshotTests.m]: Tip: I would suggest snapshotting the actionsheet's view instead. In the past when I attempted to snapshot a Window (for Dialogs) it caused a crash in a different test.

**Review by romoore — CHANGES_REQUESTED** [review]
I think the internal fixes should be made separately from the Dynamic Type changes. That will help isolate any breaks caused by either change when the release lands.  If you think they need to land together, let's chat offline about it.

File: components/ActionSheet/examples/ActionSheetComparisonExampleViewController.m (hunk 1 above)
Comment (codeman7) [reply components/ActionSheet/examples/ActionSheetComparisonExampleViewController.m]: I left these here to showcase the new behavior but didn't feel strongly about it. They have been removed.

File: components/ActionSheet/src/MDCActionSheetController.m (hunk 2 above)
Comment (codeman7) [reply components/ActionSheet/src/MDCActionSheetController.m:182]: These were broken out into #8735

File: components/ActionSheet/src/private/MDCActionSheetHeaderView.h (hunk 3 above)
Comment (codeman7) [reply components/ActionSheet/src/private/MDCActionSheetHeaderView.h]: I've updated the header to conform to the protocol and reverted these changes. Please take another look here.

File: components/ActionSheet/src/private/MDCActionSheetItemTableViewCell.h (hunk 4 above)
Comment (codeman7) [reply components/ActionSheet/src/private/MDCActionSheetItemTableViewCell.h]: I've updated the cell to conform to the protocol and reverted these changes. Please take another look here.

File: components/ActionSheet/tests/snapshot/MDCActionSheetControllerSnapshotTests.m (hunk 5 above)
Comment (codeman7) [reply components/ActionSheet/tests/snapshot/MDCActionSheetControllerSnapshotTests.m]: Done

File: components/ActionSheet/src/private/MDCActionSheetHeaderView.h (hunk 3 above)
Comment (romoore) [reply components/ActionSheet/src/private/MDCActionSheetHeaderView.h]: Ack.

File: components/ActionSheet/src/private/MDCActionSheetItemTableViewCell.h (hunk 4 above)
Comment (romoore) [reply components/ActionSheet/src/private/MDCActionSheetItemTableViewCell.h]: Ack.

File: components/ActionSheet/src/MDCActionSheetController.m [hunk 6]
```diff
@@ -423,6 +438,14 @@ - (void)setShowsHeaderDivider:(BOOL)showsHeaderDivider {
 
 #pragma mark - Dynamic Type
 
+- (void)setAdjustsFontForContentSizeCategory:(BOOL)adjustsFontForContentSizeCategory {
+  if (@available(iOS 10.0, *)) {
+    _adjustsFontForContentSizeCategory = adjustsFontForContentSizeCategory;
+    self.header.adjustsFontForContentSizeCategory = _adjustsFontForContentSizeCategory;
+    [self updateTable];
```
Comment (romoore) [inline components/ActionSheet/src/MDCActionSheetController.m:445]: Tip: Does the change in fonts require a full reload of data, or is it sufficient to just call `setNeedsLayout` on  `tableView`?

File: components/ActionSheet/src/private/MDCActionSheetItemTableViewCell.m [hunk 7]
```diff
@@ -138,6 +138,9 @@ - (MDCActionSheetAction *)action {
 }
 
 - (void)setActionFont:(UIFont *)actionFont {
+  if ([actionFont isEqual:_actionFont]) {
```
Comment (romoore) [inline components/ActionSheet/src/private/MDCActionSheetItemTableViewCell.m:141]: This looks like it might break support for `MDCFontLoader` fonts since it doesn't check if `mdc_scalingCurve` is equivalent (or set). 

Is there a way to avoid reassigning fonts when the trait collection changes that doesn't require this check in `setActionFont:`?

File: components/ActionSheet/src/private/MDCActionSheetItemTableViewCell.m [hunk 8]
```diff
@@ -155,7 +158,18 @@ - (void)updateTitleFont {
...
+- (void)setAdjustsFontForContentSizeCategory:(BOOL)adjustsFontForContentSizeCategory {
+  self.actionLabel.adjustsFontForContentSizeCategory = adjustsFontForContentSizeCategory;
+}
+
+- (BOOL)adjustsFontForContentSizeCategory {
+  return self.actionLabel.adjustsFontForContentSizeCategory;
+}
+
 - (void)mdc_setAdjustsFontForContentSizeCategory:(BOOL)adjusts {
+  if (_mdc_adjustsFontForContentSizeCategory == adjusts) {
```
Comment (romoore) [inline components/ActionSheet/src/private/MDCActionSheetItemTableViewCell.m:170]: Is this required for the UIContentSizeCategoryAdjusting behavior or can it be split to another PR?

File: components/ActionSheet/tests/snapshot/MDCActionSheetControllerSnapshotTests.m [hunk 9]
```diff
@@ -968,4 +984,105 @@ - (void)testActionSheetWithHeaderShownButNoTitleOrMessage {
...
+    UIFontMetrics *messageMetrics = [UIFontMetrics metricsForTextStyle:UIFontTextStyleSubheadline];
+    UIFont *messageFont = [messageMetrics scaledFontForFont:originalFont
+                              compatibleWithTraitCollection:extraSmallTraits];
+    self.actionSheetController.messageFont = messageFont;
+    // Setup actions font
+    UIFontMetrics *bodyMetrics = [UIFontMetrics metricsForTextStyle:UIFontTextStyleBody];
+    UIFont *bodyFont = [bodyMetrics scaledFontForFont:originalFont
+                        compatibleWithTraitCollection:extraSmallTraits];
+    self.actionSheetController.actionFont = bodyFont;
+    self.actionSheetController.view.bounds = CGRectMake(0, 0, 1000, 1000);
```
Comment (romoore) [inline components/ActionSheet/tests/snapshot/MDCActionSheetControllerSnapshotTests.m:1070]: Tip: Is there any reason why these snapshots need to be 1000x1000? Can you not use the preferredContentSize or something similar to size the view?

**Review by romoore — CHANGES_REQUESTED** [review]
I'm worried some of the optimizations around font setting might break MDCFontLoader-based fonts. Perhaps we can chat offline tomorrow about the need for the equality check?

**Comment by codeman7** [comment]
Not currently working on this might come back to it at a later date.
